[v11] DatabaseService heartbeat backport

marcoandredinis opened the pull request:
This PR backports the initial PR that added DatabaseService heartbeat.
https://github.com/gravitational/teleport/pull/19940
https://github.com/gravitational/teleport/pull/19453

This PR introduced two bugs:
- Multiple versions within the same cluster:
https://github.com/gravitational/teleport/issues/19919
- backward compatibility for trusted clusters
  when the root is upgraded to a new version, there's an error because the Leaf isn't aware of the new resource: DatabaseService Resources

A fix was created for these two bugs (see the 2nd commit).
Because of scenario C, we had to remove DatabaseService from the proxy watchers list.
This means that Proxies will always hit the Auth's backend when listing DatabaseServices (instead of using the cache).
Given that this resource is only used to assist a manual process (Users adding Databases using the Discover Flow), we think it's worth it to keep backwards and forwards (for all the v11.x.y releases) compatible

The following tests were performed:
A)
Start Root with 11.2.2 and Leaf with 11.2.2
Upgrade the Root cluster to this branch (11.x)
Upgrade the Root to v12 build
There was no error during the upgrades and Database Services were available in Root after the first upgrade.
:heavy_check_mark: 

B)
Single cluster
Start 3 services: Auth, one Proxy and one DB Service with 11.2.2

Upgrade Auth to 11.x, check for errors
Upgrade Proxy to 11.x, check for errors, downgrade to 11.2.2
Upgrade DB Service to 11.x, check for errors
Upgrade Proxy to 11.x

Upgrade Auth to 12, check for errors.
Upgrade Proxy to 12, check for errors, downgrade to 11.x
Upgrade DB Service to 12, check for errors
Upgrade Proxy to 12

✅

C)
Single cluster with Auth service having a version prior to Proxy.
After removing the `DatabaseService` kind from the `ForProxy` watchers, the following was tested:

Start 3 services: Auth, one Proxy and one DB Service with 11.2.2
Upgrade Proxy to 11.x, check for errors, downgrade to 11.2.2
Upgrade DB Service to 11.x, check for errors
Upgrade Proxy to 11.x, check for errors
Upgrade Auth to 11.x, check for errors

Upgrade Proxy to 11.x -> returns an error: Auth must be upgraded first
Upgrade Auth to 12, check for errors.
Upgrade Proxy to 12, check for errors, downgrade to 11.x
Upgrade DB Service to 12, check for errors
Upgrade Proxy to 12

✅

marcoandredinis committed:
Fix DatabaseService cache configs for old remote proxies.

Add failsafe checks when the remote proxy doesn't know about
DatabaseServices resource.
This can happen when the remote proxy has a version prior to this commit
(currently this would be v11.2.2).

marcoandredinis committed:
Revert "Revert "[v11] DatabaseService: CRUD and hearbeat (#19453)" (#19940)"

This reverts commit a82d7fece40a75fd61831ae03f5215047da6510e.

marcoandredinis committed:
Remove watcher DB Service in the ForProxy

codingllama commented on lib/srv/heartbeat.go:
```diff
@@ -571,6 +571,11 @@ func (h *Heartbeat) announce() error {
 			keepAlive, err := h.Announcer.UpsertDatabaseService(h.cancelCtx, dbService)
 			if err != nil {
+				// When Auth Service is running on a previous minor/patch version than 11.2.2, it doesn't know about DatabaseServices
+				// In this case there's not heartbeat.
```
nit:

```suggestion
				// In this case there isn't a heartbeat.
```

codingllama reviewed (approved):
LGTM.

Can't say I've gone line-by-line in the big backport commit, but I did check the others. I suggest tagging someone else from the DB team to double-check the heartbeat and Proxy watcher changes.

Apologies for the delay in reviewing.

marcoandredinis committed:
fix typo

smallinsky reviewed (approved):
It is quite big backport PR so it is easy to miss something  going through backport changes but if all tests from PR description was successfully preformed then it look fine for me.

marcoandredinis commented:
@smallinsky @codingllama Thank you for the review 🙏
If you know more scenarios worth of testing, please let me know.
I'm totally onboard on being extra cautious here

r0mant commented on lib/services/local/presence.go:
```diff
@@ -1577,6 +1580,38 @@ func (s *PresenceService) listResources(ctx context.Context, req proto.ListResou
 }
 
+// UpsertDatabaseService creates or updates (by name) a DatabaseService resource.
+func (s *PresenceService) UpsertDatabaseService(ctx context.Context, service types.DatabaseService) (*types.KeepAlive, error) {
```
Why is this method not in lib/services/local/databaseservice.go? Other database service backend methods are there.

marcoandredinis replied:
I tried to recall the reasoning on this one, but I don't recall it.

Well, I think this should be changed.
Given that this PR is a backport, I would rather open a new PR against master and backport it to v12 and v11

Unless, you think this might cause any issues when running.

r0mant replied:
Up to you, new PR to move it sounds good to me too.

marcoandredinis replied:
Created here
https://github.com/gravitational/teleport/pull/20930

marcoandredinis replied:
The above PR was merged into `master` and I'm also including it here in this PR (cherry pick from `master`)

r0mant commented on lib/services/presets.go:
```diff
@@ -75,6 +75,9 @@ func NewPresetEditorRole() types.Role {
 					types.NewRule(types.KindInstaller, RW()),
 					types.NewRule(types.KindInstance, RO()),
 					types.NewRule(types.KindLoginRule, RW()),
+					types.NewRule(types.KindLicense, RO()),
+					types.NewRule(types.KindDownload, RO()),
```
Why are these 2 needed?

marcoandredinis replied:
Oops
They shouldn't be here, removing
I think they came from a merge conflict

r0mant commented on lib/services/presets.go:
```diff
@@ -179,6 +182,7 @@ func defaultAllowRules() map[string][]types.Rule {
 			types.NewRule(types.KindConnectionDiagnostic, RW()),
 			types.NewRule(types.KindDatabase, RW()),
 			types.NewRule(types.KindLoginRule, RW()),
+			types.NewRule(types.KindDatabaseService, RO()),
```
Should KindLicense and KindDownload be included here as well? You added them to the editor role above.

marcoandredinis replied:
No, removed them in the editor role.
Thank you

r0mant commented on lib/services/role.go:
```diff
@@ -2015,10 +2015,16 @@ func (set RoleSet) checkAccess(r AccessCheckable, mfa AccessMFAParams, matchers
 		getRoleLabels = types.Role.GetDatabaseLabels
 		additionalDeniedMessage = "Confirm database user and name."
+	case types.KindDatabaseService:
+		resourceHasLabels = false
+		getRoleLabels = getEmptyLabels
```
This is a little confusing tbh. Why the special case for database service?

marcoandredinis replied:
I agree, it's not very clear.

In order to re-use the ListResources method for DatabaseServices, we end here.
The other resources have associated labels in the role
KindDatabase: `db_labels`
KindApp: `app_labels`
KindNode: 'node_labels`
And so on.

However, DatabaseServices doesn't have an associated `<x>_labels` field in the Role.Allow spec.
So, to be able to re-use the ListResources, I had to add a little branch on the code to allow for that case.

There was a little discussion about this in the `master`'s PR
https://github.com/gravitational/teleport/pull/19469#discussion_r1052271379

r0mant replied:
@marcoandredinis I guess my point is, why not just add `GetDatabaseServiceLabels` to the new resource and use it here for consistency without introducing extra flags, even it if will always return empty? If you look below, WindowsDesktop/WindowsDesktopService do this, basically similar situation, no?

I'd just really like to avoid extra scenarios in RBAC checks, opens up a way for errors.

marcoandredinis replied:
The WindowsDesktop and WindowsDesktopService use the `windows_desktop_labels` set of labels.
So it's actually treated as a single resource. You cannot set different labels for WindowsDesktop and WindowsDesktopService.

I'm not sure about grouping applying the same logic here: have the `db_labels` as the source for the DatabaseServices allowed labels.
I see them as two separate resources.

Is your suggestion to use `db_labels` as the source for allowing access to DatabaseServices?
Or should we create a `GetDatabaseServiceLabels` that always returns `'*': '*'` ? ( I don't think returning an empty set of labels works)

marcoandredinis replied:
After talking to Roman we concluded that this is hackish at best and will instead create a `db_services_labels` as Role Allow/Deny Option and set it to 
A new PR will be created and once approved, cherry-pick it to this backport PR

r0mant commented on lib/srv/db/server.go:
```diff
@@ -615,6 +615,11 @@ func (s *Server) Start(ctx context.Context) (err error) {
 	// Start cloud users that will be monitoring cloud users.
 	go s.cfg.CloudUsers.Start(ctx, s.getProxiedDatabases)
 
+	// Start hearbeating the Database Service itself.
```
```suggestion
	// Start heartbeating the Database Service itself.
```

r0mant commented on lib/srv/db/server.go:
```diff
@@ -642,13 +647,56 @@ func (s *Server) Start(ctx context.Context) (err error) {
 
 	// If the agent doesn’t have any static databases configured, send a
 	// heartbeat without error to make the component “ready”.
+	// This is not necessary on next versions (v12+) because they have a Heartbeat for the Service itself.
```
```suggestion
	// DELETE IN 13.0: This is not necessary on next versions (v12+) because they have a Heartbeat for the Service itself.
```

r0mant commented on lib/srv/heartbeat.go:
```diff
@@ -133,6 +135,8 @@ const (
 	HeartbeatModeApp
 	// HeartbeatModeDB sets heatbeat to db
 	HeartbeatModeDB
+	// HeartbeatModeDatabaseService sets heatbeat mode to DatabaseService.
```
```suggestion
	// HeartbeatModeDatabaseService sets heartbeat mode to DatabaseService.
```

r0mant commented on lib/srv/heartbeat.go:
```diff
@@ -560,6 +564,32 @@ func (h *Heartbeat) announce() error {
+				// When Auth Service is running on a previous minor/patch version than 11.2.2, it doesn't know about DatabaseServices
+				// In this case there isn't a heartbeat.
+				if trace.IsNotImplemented(err) {
+					return nil
```
Nit: Log something? E.g. `log.Debug("Auth server doesn't recognize DatabaseService resource. Upgrade auth server.")`.

marcoandredinis committed:
improve logs and fix typos

marcoandredinis committed:
Simplify role access checker

marcoandredinis committed:
DatabaseService: move Upsert op to DBServicesService (#20930)

marcoandredinis committed:
Add RBAC labels for Database Services access (#21093)

* Add RBAC labels for Databas Services access

This PR adds a new set of Labels that controls access to the
DatabaseService resource.

We changed the `RoleSet.checkAccess` to allow for checks in Resources
without Labels.
However adding complexity to this function is something that can bit us
in the future.
So, this PR reverts that change and adds a new set of labels.
Those labels will control whether the RoleSet has access to
DatabaseServices.

Given that this is a new condition that we add to the default `editor`
role, we had to create a migration/backfill for the roles that don't
have them.
This allow us to migrate the roles created previously and add this new
allow condition.

* remove preset for remote proxy and database labels

* Remove db_service_labels wildcard from editor preset role

* Fix early exit logic
